src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java

```diff
@@ -2920,41 +2930,75 @@ private void fireLineWidgetsChanged()
       AceEditor.this.fireEvent(new LineWidgetsChangedEvent());
    }
    
-   private static class ScopeTimer
+   private static class BackgroundTokenizer
    {
-      public ScopeTimer(final AceEditor editor)
+      public BackgroundTokenizer(final AceEditor editor)
       {
+         editor_ = editor;
+         
          timer_ = new Timer()
          {
             @Override
             public void run()
             {
-               if (row_ >= editor.getRowCount())
+               // Stop our timer if we've tokenized up to the end of the document.
+               if (row_ >= editor_.getRowCount())
+               {
+                  // If the timer was triggered by a document change event,
+                  // then fire an event signaling that we have a new scope
+                  // tree ready.
+                  if (mutated_)
+                  {
+                     editor_.fireEvent(new ScopeTreeReadyEvent(
+                           editor_.getScopeTree(),
+                           editor_.getCurrentScope()));
+                     mutated_ = false;
+                  }
                   return;
+               }
                
                row_ += ROWS_TOKENIZED_PER_ITERATION;
-               row_ = editor.buildScopeTreeUpToRow(row_);
+               row_ = Math.max(row_, editor.buildScopeTreeUpToRow(row_));
                timer_.schedule(DELAY_MS);
             }
          };
          
-         editor.addCursorChangedHandler(new CursorChangedHandler()
+         editor_.addDocumentChangedHandler(new DocumentChangedEvent.Handler()
+         {
+            @Override
+            public void onDocumentChanged(DocumentChangedEvent event)
+            {
+               mutated_ = true;
+               row_ = event.getEvent().getRange().getStart().getRow();
+               timer_.schedule(editor_.getSuggestedScopeUpdateDelay());
+            }
+         });
+         
+         // We signal on cursor change as well so that the background
+         // tokenization pauses while the user is scrolling or typing.
+         editor_.addCursorChangedHandler(new CursorChangedHandler()
          {
             @Override
             public void onCursorChanged(CursorChangedEvent event)
             {
-               int row = event.getPosition().getRow();
-               row_ = row;
-               timer_.schedule(editor.getSuggestedScopeUpdateDelay() / 2);
+               timer_.schedule(editor_.getSuggestedScopeUpdateDelay());
             }
          });
       }
       
+      public boolean isReady(int row)
+      {
+         return row < row_;
+      }
+      
+      private final AceEditor editor_;
       private final Timer timer_;
+      
       private int row_ = 0;
+      private boolean mutated_ = true;
       
-      private static final int DELAY_MS = 5;
-      private static final int ROWS_TOKENIZED_PER_ITERATION = 50;
+      private static final int DELAY_MS = 0;
```

Comment:
Note that this change causes much more eager background tokenization, but the background tokenizer stops as soon as the user starts moving around in the editor (cursor movements etc.).

I've been tweaking this as 100 just seemed too slow in larger documents and didn't confer the benefits we'd hope to get from background tokenization. The downside is that if the user starts editing while the background tokenizer is running in a large document, there will be a brief ~500ms pause until that particular iteration finishes.